jurity/recommenders/interlist_diversity.py

```diff
@@ -0,0 +1,225 @@
+                 sample_size: float = None, seed: int = Constants.default_seed) -> pd.DataFrame:
+    """
+    Samples input data frame by selecting a random sample of users.
+
+    Parameters
+    ----------
+    df: pd.DataFrame
+        Data frame with a user_id_col column.
+    user_id_column: str
+        User id column name.
+    sample_size: float
+        Proportion of users to randomly sample for evaluation.
+    seed : int, default=Constants.default_seed
+        The seed used to create random state.
+
+    Returns
+    -------
+    Sampled data frame
+    """
+    rng = np.random.RandomState(seed)
+    users = df[user_id_column].unique()
+    users = rng.choice(users, size=int(len(users) * sample_size), replace=False)
+    return df[df[user_id_column].isin(users)]
+
+
+def tocsr(df: pd.DataFrame, user_id_column: str = Constants.user_id, item_id_column: str = Constants.item_id):
+    """
+    Transform data frame with user_id and item_id columns to sparse csr matrix.
+
+    Parameters
+    ----------
+    df: pd.DataFrame
+        Data frame with (user_id, item_id) in each row.
+    user_id_column: str
+        User id column name.
+    item_id_column: str
+        Item id column name.
+
+    Returns
+    -------
+    Sparse matrix with user-item interactions/recommendations.
+    """
+
+    # Map each user_id to (0, n_users)
+    users = list(df[user_id_column].unique())
+    user_map = dict(zip(users, range(len(users))))
+
+    # Map each user_id to (0, n_items)
+    items = list(df[item_id_column].unique())
+    item_map = dict(zip(items, range(len(items))))
+
+    # Update user_id, item_id values
+    user_ids = df[user_id_column].map(user_map).values
+    item_ids = df[item_id_column].map(item_map).values
+
+    # Convert to csr matrix
+    csr_matrix = sp.coo_matrix((np.ones(len(user_ids)), (user_ids, item_ids))).tocsr()
+
+    return csr_matrix
+
+
+def interlist_diversity(predicted_results: pd.DataFrame, click_column: str, k: int,
+                        user_id_column: str = Constants.user_id, item_id_column: str = Constants.item_id,
+                        sample_size: float = None, seed: int = Constants.default_seed,
+                        chunk_size: int = 1000) -> Tuple[float, int]:
+    """
+        Calculate inter-list diversity metric:
+
+            Inter-list-diversity = 1 - average(cosine_similarity(R_{u_i}, R_{u_j})),
+            where R_{u_i} is the binary indicator vector representing provided recommendations for user u_i, i \ne j.
+
+        Parameters
+        ----------
+        predicted_results: pd.DataFrame
+            Recommendations data frame with (user_id, item_id, score) in each row.
+        k: int
+            Top-k recommendations to consider.
+        user_id_column: str
+            User id column name.
+        item_id_column: str
+            Item id column name.
+        click_column: str
+            Recommendation score column name.
+        sample_size: float
```

Comment:
I am not sold on "sample_size". 

Naming: This is not size, but really a "user_sample_ratio" as it is implemented here. 

Behavior: Maybe it should really "user_sample_size" instead of a ratio. 

The default behavior here is None, that is to use all data. If I remember correctly from our RecWiser experience, this operation can bog down quickly as it is quadratic in the number of user X items, which can get intractable in most practical scenarios.  

We might not want to end up with situations/issues where every user trying Jurity (or the RecWiser notebooks eventually) to get back to us with "oh but this never finishes.."

Should we make the default behavior an approx solution instead? Some default we know will finish and give some results. This might be preferred especially when benchmarking multiple algos as in RecWiser where we are only interested in finding the top 2-3 algos that work well --rather than precisely knowing the diversity measure of the 10th worst method. 

If so, what that default should be? I don't know if we can make this "ratio" anymore since we don't know data size it could be used for. But instead, we might say, a random sample of 1000, or 10K?, etc users. If there are fewer users than that, fine, we have an exact solution but otherwise, we get an approximation. 

Btw, if the default behavior is gonna be an approximation, should we consider a "num_run" parameter to repeat this k-many times and return the average of the sample as the result for robustness?

Reply:
I really like the idea of using an approximate solution as a default. Especially some fresh user coming in would not necessarily notice the O(n2) complexity and would become frustrated. By extension, this default approximate solution could be a feature request for Selective where possible (speaking from experience :D )

Reply:
Agree that it would be easy to reach the limit of computing capacity while dealing with entire users X items. Keeping this running for a long time would be unwanted, especially during the benchmarking across varying setups. Set approx solution as default behavior with averaged over `num_run` runs makes sense. We keep the capacity to run on the entire data when data is not large or if it is really needed.

Reply:
Cool! I don't what would be a reasonable "user_sample_size" be, 100? 1000? 10000? but I am sure you guys have some experience setting a sensible default that works for us.

Reply:
A few updates to "user_sample_size" include:

- Set default behavior to be a real size, default=10000. Tested with the newly added parameter `num_runs` (default=10) for getting reasonable speed-up in evaluations and close approximations. 
- `user_sample_size` is able to take both int type input and float input, with the float indicating the ratio as what it currently does.
- When `user_sample_size` is set to be larger than total number of users. Program uses the total number of users and shows a warning.

Reply:
NICE!